### ISSUE TITLE
__ESCAPED_SOURCE_END_CLEAN_CSS__

### ISSUE THREAD

**Issue by silverwind** [jakubpawlowicz/clean-css#628]
I'm seeing this string inserted at the end of my output. Unfortunately I can't easily provide a example to reproduce, but I'm wondering if you could tell me the conditions under which it is inserted. I assume it's meant to be removed from the final output?

**Comment by Mottie** [comment]
It looks like it's from here - [/lib/text/escape-store.js](https://github.com/jakubpawlowicz/clean-css/blob/master/lib/text/escape-store.js#L4)

**Comment by jakubpawlowicz** [comment]
Basically speaking it's a marker for every file that is processed by clean-css, so we can track sources properly. 

In fact it's added by SourceTracker object, which is instantiated here - https://github.com/jakubpawlowicz/clean-css/blob/d21f1e0808dfc1a658f745cad93e8b74466ffe73/lib/clean.js#L78

It should be removed when serializing the tokenized source. Let's find out why it's not.

**Comment by silverwind** [comment]
Will try to provide a reduced test case. Maybe the issue is related to special comments. Looking at the [removal regex](https://github.com/jakubpawlowicz/clean-css/blob/4fed29d8989494ef423163bacfe75d0472a6ff0a/lib/utils/source-tracker.js#L25-L29), I'm wondering why it isn't multiline (flags `gm`).

**Comment by Mottie** [comment]
Ok, I have a reduced test case, and I don't think it's clean-css' fault.

If we start out with this css:

```css
pre,
/*[[syntax-theme]]*/
/* grunt build - remove to end of file */
.highlight {
  background: #141414 !important;
  color: #ddd !important;
}
```

Our grunt build removes everything after the grunt build comment. So we end up with this after css minification

```css
pre,
/*[[syntax-theme]]*/
__ESCAPED_SOURCE_END_CLEAN_CSS__
}
```

The reason we have a trailing comma is because this version is posted to userstyles and the `/*[[syntax-theme]]*/` is replaced by user selected styles. We include the selector before before that code because it is easier to modify selectors in one place than in every syntax theme.

**Comment by silverwind** [comment]
Indeed the issue is with our invalid CSS. We shall fix that.

@jakubpawlowicz here's a minimal example to reproduce:

```
$ cat x.css
.x,/*!*/
$ cleancss x.css
.x,/*!*/
__ESCAPED_SOURCE_END_CLEAN_CSS__%
```

**Comment by jakubpawlowicz** [comment]
Thanks guys for tracking it down. 

@silverwind - I'll look into handling that case too.

**Comment by madwizard-thomas** [comment]
The token also appears in warnings or output when a closing bracket is missing. Calling cleancss on a single file with contents
**a.css**
```
div{color:blue;
```
will show:
```
$ cleancss a.css
WARNING: Missing '}' after 'color:blue;__ESCAPED_SOURCE_END_CLEAN_CSS__'. Ignoring.
```
Now add **b.css**
```
p{background:red;}
```
Result:
```
$ cleancss a.css b.css
div{color:#00f}
```
Note that the p tag from b.css is missing.

These bugs are caused by the whatsNext function in the tokenizer (lib/tokenizer/tokenize.js). As soon as it is in the body context it [doesn't check](https://github.com/jakubpawlowicz/clean-css/blob/a00f5d6ae61b79a83d86b04aaf865f1ad64f6240/lib/tokenizer/tokenize.js#L68-L77) for any escape tokens anymore but only looks for the nearest closing bracket (which because of the concatenation may be from a different file).

**Comment by jakubpawlowicz** [comment]
Aha, thanks @madwizard-thomas for investigating it! 

I'm in the process of preparing a new clean-css version which moves away from escaping content, this this bug will be no more. Given a fix is easy we should still patch the 3.4 branch though.

**Comment by jakubpawlowicz** [comment]
This bug is fixed on master now.
